WordPress/gutenberg
# Improve Nav block loading and placeholder states

## Description
Currently the loading and placeholder states for the Nav block are quite difficult to reason about. 

- There's a lot of component nesting and states within states. An example is the `PlacholderPreview` which is rendered at both the root block level _and_ within the `PlaceholderComponent` itself. 
- The logic determining when the placeholder is shown requires setting state when this can simply be derived from existing state.
- The loading states are too complex and obscured.
- The loading state visuals are difficult to distinguish.
- Little or no non-visual communication of loading progress.

This PR seeks to solve the above issues by:

- deriving the placeholder visibility from existing state rather than requiring `setState()`s.
- add a clear "loading" spinner to the loading state as discussed with @jasmussen.
- consolidate all loading state logic into the block _root_ - leaving only "locally relevant" loading logic (i.e. resolution status of which optoins are available within the placeholder actions) within the Placeholder component.
- avoid potential for overlapping rendering by returning distinct components early for both placeholder and loading states
- adds `aria-live` announcement of block loading progress.
- avoid rendering both placeholder preview and block appender when the selected menu is empty

🙏 **Please note**:  this PR does not attempt to fix the "freezing" of the UI that occurs when you create a new Menu or select an existing one. That issue is addressed in https://github.com/WordPress/gutenberg/pull/38858. The aim of this PR is simply to iterate towards an improved loading state for the block.

## Testing Instructions





I recommend testing the scenario on `trunk` first to get a sense of just how suboptimal the current loading experience is.

Now on this PR's branch:

1. Create new Post.
2. Set Devtools to `Slow 3G` and add a Nav block.
3. Notice the loading spinner.
4. Choose an existing Menu with items.
5. Save and reload page.
6. See loading spinner whilst items are being resolved.
7. Add a new Nav block and choose create _empty_.
8. Save and reload page.
9. See loading state whilst items are resolved and then see the placeholder (this shows because there are no items in the menu/block).

## Screenshots 

#### Before

No clear loading state for initial setup or subsequent loading of existing blocks:

https://user-images.githubusercontent.com/444434/154667604-529a9879-2394-42b1-a949-faa8361f6adf.mov


#### After

Clear loading state for initial setup _and_ subsequent loading of existing blocks:


https://user-images.githubusercontent.com/444434/155123560-cd7a43b4-e1f2-4d2f-893e-f8ff31d1ae41.mov


## Types of changes





## Checklist:
